pkg/gardenlet/controller/managedseed/charttest/charttest.go

```diff
@@ -1015,64 +1015,96 @@ func ComputeExpectedGardenletDeploymentSpec(
-						{
-							LabelSelector: &metav1.LabelSelector{
-								MatchExpressions: []metav1.LabelSelectorRequirement{
-									{
-										Key:      "app",
-										Operator: "In",
-										Values:   []string{"gardener"},
-									},
-									{
-										Key:      "role",
-										Operator: "In",
-										Values:   []string{"gardenlet"},
-									},
-								},
-							},
-							TopologyKey: "kubernetes.io/hostname",
+		}
+
+		topologySpreadConstraintLabels := map[string]string{
+			"app":      "gardener",
+			"role":     "gardenlet",
+			"chart":    "runtime-0.1.0",
+			"release":  "gardenlet",
+			"heritage": "Tiller",
+		}
+
+		if failureToleranceType := deploymentConfiguration.FailureToleranceType; failureToleranceType != nil {
+			if *failureToleranceType == gardencorev1beta1.FailureToleranceTypeNode {
+				deployment.Template.Spec.TopologySpreadConstraints = []corev1.TopologySpreadConstraint{
+					{
+						MaxSkew:     1,
+						TopologyKey: "kubernetes.io/hostname",
+						LabelSelector: &metav1.LabelSelector{
+							MatchLabels: topologySpreadConstraintLabels,
 						},
+						WhenUnsatisfiable: corev1.DoNotSchedule,
 					},
-				},
+				}
 			}
-		}
 
-		if deploymentConfiguration.Env != nil {
-			deployment.Template.Spec.Containers[0].Env = deploymentConfiguration.Env
+			if *failureToleranceType == gardencorev1beta1.FailureToleranceTypeZone {
+				deployment.Template.Spec.TopologySpreadConstraints = []corev1.TopologySpreadConstraint{
+					{
+						MaxSkew:     1,
+						TopologyKey: "kubernetes.io/hostname",
+						LabelSelector: &metav1.LabelSelector{
+							MatchLabels: topologySpreadConstraintLabels,
+						},
+						WhenUnsatisfiable: corev1.DoNotSchedule,
+					},
+					{
+						MaxSkew:     1,
+						TopologyKey: "topology.kubernetes.io/zone",
+						LabelSelector: &metav1.LabelSelector{
+							MatchLabels: topologySpreadConstraintLabels,
+						},
+						WhenUnsatisfiable: corev1.DoNotSchedule,
+					},
+				}
+			}
 		}
 
-		if deploymentConfiguration.PodLabels != nil {
-			deployment.Template.ObjectMeta.Labels = utils.MergeStringMaps(deployment.Template.ObjectMeta.Labels, deploymentConfiguration.PodLabels)
+		if pointer.Int32Deref(deployment.Replicas, 1) > 1 && deployment.Template.Spec.TopologySpreadConstraints == nil {
+			deployment.Template.Spec.TopologySpreadConstraints = append(deployment.Template.Spec.TopologySpreadConstraints, corev1.TopologySpreadConstraint{
+				MaxSkew:     1,
+				TopologyKey: "kubernetes.io/hostname",
+				LabelSelector: &metav1.LabelSelector{
+					MatchLabels: topologySpreadConstraintLabels,
+				},
+				WhenUnsatisfiable: corev1.ScheduleAnyway,
+			})
 		}
+	}
 
-		if deploymentConfiguration.PodAnnotations != nil {
-			deployment.Template.ObjectMeta.Annotations = utils.MergeStringMaps(deployment.Template.ObjectMeta.Annotations, deploymentConfiguration.PodAnnotations)
-		}
+	if deploymentConfiguration.Env != nil {
```

Comment:
This and the other checks below were moved so that they are now outside the block that checks 
`if deploymentConfiguration != nil {` which could cause nil pointer exceptions.